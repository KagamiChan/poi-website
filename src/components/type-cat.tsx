import classNames from 'classnames'
import React from 'react'
import Typist from 'react-typist'

import styles from './type-cat.module.css'

interface IProps {
  text: string
}

interface IState {
  stage: string
  width: number
}

export class TypeCat extends React.Component<IProps, IState> {
  public state = {
    stage: 'loaded',
    width: 0,
  }

  private mirror = React.createRef<HTMLDivElement>()

  public componentDidMount() {
    this.setState({
      stage: 'typing',
      width: this.mirror.current!.getBoundingClientRect().width,
    })
  }

  public handleTypingDone = () => {
    setTimeout(() => {
      this.setState({
        stage: 'done',
      })
    }, 1000)
  }

  public render() {
    return (
      <div className={styles.container}>
        <div
          className={classNames({
            [styles.hide]: this.state.stage !== 'typing',
          })}
          style={{ width: this.state.width }}
        >
          <Typist
            cursor={{
              element: '_',
              hideWhenDone: true,
            }}
            onTypingDone={this.handleTypingDone}
          >
            {this.props.text}
          </Typist>
        </div>
        <div
          ref={this.mirror}
          className={classNames({
            [styles.mirror]: true,
            [styles.hide]: this.state.stage === 'typing',
            [styles.show]: this.state.stage === 'done',
          })}
          data-testid={this.state.stage}
        >
          {this.props.text}
        </div>
      </div>
    )
  }
}
