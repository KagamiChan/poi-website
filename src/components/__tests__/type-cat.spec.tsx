import { render, wait } from '@testing-library/react'
import React from 'react'

import { TypeCat } from '../type-cat'

describe('<TypeCat />', () => {
  it('renders', () => {
    const { asFragment } = render(<TypeCat text="Hello World" />)
    expect(asFragment()).toMatchSnapshot()
  })
})
