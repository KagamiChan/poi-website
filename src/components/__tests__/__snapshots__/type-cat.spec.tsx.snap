// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TypeCat /> renders 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class=""
      style="width: 0px;"
    >
      <div
        class="Typist "
      >
        <span
          class="Cursor Cursor--blinking"
        >
          _
        </span>
      </div>
    </div>
    <div
      class="mirror hide"
      data-testid="typing"
    >
      Hello World
    </div>
  </div>
</DocumentFragment>
`;
